polyfill.d.ts

```diff
@@ -0,0 +1,21 @@
+interface ReadonlyArray<T> {
+  flatMap<U, This = undefined> (callback: (this: This, value: T, index: number, array: T[]) => U|U[], thisArg: This = undefined): U[]
+
+  flatten (depth: 0): T[]
+  flatten<U extends any[]> (this: U[], depth?: 1 = 1): U
+  flatten<U extends any[]> (this: U[][], depth: 2): U
+  flatten<U extends any[]> (this: U[][][], depth: 3): U
+  flatten<U extends any[]> (this: U[][][][], depth: 4): U
+  flatten<U> (depth?: number = 1): U[]
```

Comment:
Interesting. I created a polyfill for this, although I will be happy to use this official one instead 
:heart:
My type declaration for flatten is quite different.

It looks like
```ts
interface Array<T> {
  flatten<T>(this: (T | T[] | this)[], depth?: number): T[];
}
```
The recursive `(this)[]` works at least 9 levels deep. But I should probably test with more. https://github.com/aluanhaddad/flat-map/blob/master/test/flat-map.ts